src/object/cli_shard.c

```diff
@@ -1203,6 +1200,7 @@ dc_obj_shard_punch(struct dc_obj_shard *shard, enum obj_rpc_opc opc,
 
 out_req:
 	crt_req_decref(req);
+	crt_req_decref(req);
```

Comment:
In all places when we pass rpc in cb_args the reference is incremented. So, for error case we should decrement it twice to destroy this rpc. In good path this reference is decremented inside the callback.

Reply:
hmm different people wrote different RPC cases so may with a few detailed difference.
For this case of dc_obj_shard_punch(), i checked again that in above L1144 take 1st ref when create the req, and L1156 take 2nd ref by addref the req. So you are right need two decref for failure handling in "out_req". (actually need not crt_req_addref in L1148, so only one decref needed for "out_req").

But for daos_rpc_send()'s failure case, should not goto out_req again, because daos_rpc_send() can handle the failure by itself.
And in common case of all success, we will leak one ref for the RPC req, because daos_rpc_send() then after complete cb called, cart only decref one in crt_rpc_complete().
So looks to me that we need to remove both L1148's crt_req_addref() and the added L1190's decref. How do you think?
Seems that we need to carefully check the RPC refcount in other cases.

Reply:
I did this carefully check and provide this patch. It works according to extensive testing. I'm not in favor to start those experiments again and play with changing reference counting.